File: client/src/test/scala/stasis/test/specs/unit/client/model/TargetEntitySpec.scala
package stasis.test.specs.unit.client.model

import java.nio.file.Paths

import org.apache.pekko.util.ByteString

import stasis.client.model.EntityRef
import stasis.client.model.TargetEntity
import stasis.test.specs.unit.UnitSpec
import stasis.test.specs.unit.client.Fixtures
import stasis.test.specs.unit.client.ResourceHelpers.StringPath

class TargetEntitySpec extends UnitSpec {
  "A TargetEntity" should "fail if different entity types provided for current and existing metadata" in {
    an[IllegalArgumentException] should be thrownBy targetFile.copy(
      existingMetadata = Fixtures.Metadata.DirectoryOneMetadata,
      currentMetadata = Some(Fixtures.Metadata.FileOneMetadata)
    )

    an[IllegalArgumentException] should be thrownBy targetFile.copy(
      existingMetadata = Fixtures.Metadata.FileOneMetadata,
      currentMetadata = Some(Fixtures.Metadata.DirectoryOneMetadata)
    )
  }

  it should "determine if its metadata has changed" in {
    targetFileWithoutCurrentMetadata.hasChanged should be(true)
    targetFileWithCurrentMetadata.hasChanged should be(false)
    targetFileWithUpdatedCurrentGroup.hasChanged should be(true)
    targetDirectoryWithoutCurrentMetadata.hasChanged should be(true)
    targetDirectoryWithCurrentMetadata.hasChanged should be(false)
    targetDirectoryWithUpdatedCurrentGroup.hasChanged should be(true)
    targetLibraryWithoutCurrentMetadata.hasChanged should be(true)
    targetLibraryWithCurrentMetadata.hasChanged should be(false)
    targetLibraryWithUpdatedCurrentAttributes.hasChanged should be(true)
  }

  it should "determine if its content has changed" in {
    targetFileWithoutCurrentMetadata.hasContentChanged should be(true)
    targetFileWithCurrentMetadata.hasContentChanged should be(false)
    targetFileWithUpdatedCurrentSize.hasContentChanged should be(true)
    targetFileWithUpdatedCurrentChecksum.hasContentChanged should be(true)
    targetDirectoryWithoutCurrentMetadata.hasContentChanged should be(false)
    targetDirectoryWithCurrentMetadata.hasContentChanged should be(false)
    targetDirectoryWithUpdatedCurrentGroup.hasContentChanged should be(false)
    targetLibraryWithoutCurrentMetadata.hasContentChanged should be(true)
    targetLibraryWithCurrentMetadata.hasContentChanged should be(false)
    targetLibraryWithUpdatedCurrentChecksum.hasContentChanged should be(true)
  }

  it should "provide its original file path" in {
    targetFile.originalRef.key should be(targetFile.existingMetadata.path)
  }

  it should "provide its destination file path" in {
    val testDestinationPath = Paths.get("/tmp/destination")
    val originalPath = Fixtures.Metadata.FileOneMetadata.path.asPath

    val expectedPathWithDefaultStructure = Paths.get(s"$testDestinationPath/$originalPath")
    val expectedPathWithoutDefaultStructure = Paths.get(s"$testDestinationPath/${originalPath.getFileName}")

    targetFile.destinationRef should be(targetFile.originalRef)

    targetFile
      .copy(
        destination = TargetEntity.Destination.Directory(path = testDestinationPath, keepDefaultStructure = true)
      )
      .destinationRef should be(EntityRef.Filesystem(expectedPathWithDefaultStructure))

    targetFile
      .copy(
        destination = TargetEntity.Destination.Directory(path = testDestinationPath, keepDefaultStructure = false)
      )
      .destinationRef should be(EntityRef.Filesystem(expectedPathWithoutDefaultStructure))
  }

  it should "provide a filesystem destination for a library ref" in {
    val testDestinationPath = Paths.get("/tmp/destination")

    val libraryFile = targetFile.copy(ref = EntityRef.Library(scheme = "photos", path = "/album/img.heic"))

    libraryFile
      .copy(
        destination = TargetEntity.Destination.Directory(path = testDestinationPath, keepDefaultStructure = false)
      )
      .destinationRef should be(EntityRef.Filesystem(Paths.get("/tmp/destination/img.heic")))
  }

  private val targetFile = TargetEntity(
    ref = Fixtures.Metadata.FileOneMetadata.path.asRef,
    destination = TargetEntity.Destination.Default,
    existingMetadata = Fixtures.Metadata.FileOneMetadata,
    currentMetadata = None
  )

  private val targetFileWithoutCurrentMetadata =
    targetFile

  private val targetFileWithCurrentMetadata =
    targetFile.copy(currentMetadata = Some(Fixtures.Metadata.FileOneMetadata))

  private val targetFileWithUpdatedCurrentGroup =
    targetFile.copy(currentMetadata = Some(Fixtures.Metadata.FileOneMetadata.copy(group = "none")))

  private val targetFileWithUpdatedCurrentSize =
    targetFile.copy(currentMetadata = Some(Fixtures.Metadata.FileOneMetadata.copy(size = 0)))

  private val targetFileWithUpdatedCurrentChecksum =
    targetFile.copy(currentMetadata = Some(Fixtures.Metadata.FileOneMetadata.copy(checksum = 0)))

  private val targetDirectory = TargetEntity(
    ref = Fixtures.Metadata.DirectoryOneMetadata.path.asRef,
    destination = TargetEntity.Destination.Default,
    existingMetadata = Fixtures.Metadata.DirectoryOneMetadata,
    currentMetadata = None
  )

  private val targetDirectoryWithoutCurrentMetadata =
    targetDirectory

  private val targetDirectoryWithCurrentMetadata =
    targetDirectory.copy(currentMetadata = Some(Fixtures.Metadata.DirectoryOneMetadata))

  private val targetDirectoryWithUpdatedCurrentGroup =
    targetDirectory.copy(currentMetadata = Some(Fixtures.Metadata.DirectoryOneMetadata.copy(group = "none")))

  private val targetLibrary = TargetEntity(
    ref = EntityRef.default(Fixtures.Metadata.LibraryOneMetadata.path),
    destination = TargetEntity.Destination.Default,
    existingMetadata = Fixtures.Metadata.LibraryOneMetadata,
    currentMetadata = None
  )

  private val targetLibraryWithoutCurrentMetadata =
    targetLibrary

  private val targetLibraryWithCurrentMetadata =
    targetLibrary.copy(currentMetadata = Some(Fixtures.Metadata.LibraryOneMetadata))

  private val targetLibraryWithUpdatedCurrentChecksum =
    targetLibrary.copy(currentMetadata = Some(Fixtures.Metadata.LibraryOneMetadata.copy(checksum = 0)))

  private val targetLibraryWithUpdatedCurrentAttributes =
    targetLibrary.copy(currentMetadata =
      Some(Fixtures.Metadata.LibraryOneMetadata.copy(attributes = ByteString("favorite=false")))
    )
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/model/TargetEntitySpec.kt
package stasis.test.client_android.lib.model

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import okio.ByteString.Companion.encodeUtf8
import stasis.client_android.lib.model.EntityRef
import stasis.client_android.lib.model.TargetEntity
import stasis.test.client_android.lib.Fixtures
import stasis.test.client_android.lib.ResourceHelpers.asRef
import stasis.test.client_android.lib.ResourceHelpers.asPath
import java.math.BigInteger
import java.nio.file.Files
import java.nio.file.Paths

class TargetEntitySpec : WordSpec({
    "A TargetEntity" should {
        val targetFile = TargetEntity(
            ref = Fixtures.Metadata.FileOneMetadata.path.asRef(),
            destination = TargetEntity.Destination.Default,
            existingMetadata = Fixtures.Metadata.FileOneMetadata,
            currentMetadata = null
        )

        val targetFileWithoutCurrentMetadata =
            targetFile

        val targetFileWithCurrentMetadata =
            targetFile.copy(currentMetadata = Fixtures.Metadata.FileOneMetadata)

        val targetFileWithUpdatedCurrentGroup =
            targetFile.copy(currentMetadata = Fixtures.Metadata.FileOneMetadata.copy(group = "none"))

        val targetFileWithUpdatedCurrentSize =
            targetFile.copy(currentMetadata = Fixtures.Metadata.FileOneMetadata.copy(size = 0))

        val targetFileWithUpdatedCurrentChecksum =
            targetFile.copy(currentMetadata = Fixtures.Metadata.FileOneMetadata.copy(checksum = BigInteger("0")))

        val targetDirectory = TargetEntity(
            ref = Fixtures.Metadata.DirectoryOneMetadata.path.asRef(),
            destination = TargetEntity.Destination.Default,
            existingMetadata = Fixtures.Metadata.DirectoryOneMetadata,
            currentMetadata = null
        )

        val targetDirectoryWithoutCurrentMetadata =
            targetDirectory

        val targetDirectoryWithCurrentMetadata =
            targetDirectory.copy(currentMetadata = Fixtures.Metadata.DirectoryOneMetadata)

        val targetDirectoryWithUpdatedCurrentGroup =
            targetDirectory.copy(currentMetadata = Fixtures.Metadata.DirectoryOneMetadata.copy(group = "none"))

        val targetLibrary = TargetEntity(
            ref = Fixtures.Metadata.LibraryOneMetadata.path.asRef(),
            destination = TargetEntity.Destination.Default,
            existingMetadata = Fixtures.Metadata.LibraryOneMetadata,
            currentMetadata = null
        )

        val targetLibraryWithoutCurrentMetadata =
            targetLibrary

        val targetLibraryWithCurrentMetadata =
            targetLibrary.copy(currentMetadata = Fixtures.Metadata.LibraryOneMetadata)

        val targetLibraryWithUpdatedCurrentChecksum =
            targetLibrary.copy(
                currentMetadata = Fixtures.Metadata.LibraryOneMetadata.copy(checksum = BigInteger("0"))
            )

        val targetLibraryWithUpdatedCurrentAttributes =
            targetLibrary.copy(
                currentMetadata = Fixtures.Metadata.LibraryOneMetadata.copy(attributes = "favorite=false".encodeUtf8())
            )

        "fail if different entity types provided for current and existing metadata" {
            shouldThrow<IllegalArgumentException> {
                targetFile.copy(
                    existingMetadata = Fixtures.Metadata.DirectoryOneMetadata,
                    currentMetadata = Fixtures.Metadata.FileOneMetadata
                )
            }

            shouldThrow<IllegalArgumentException> {
                targetFile.copy(
                    existingMetadata = Fixtures.Metadata.FileOneMetadata,
                    currentMetadata = Fixtures.Metadata.DirectoryOneMetadata
                )
            }
        }

        "determine if its metadata has changed" {
            targetFileWithoutCurrentMetadata.hasChanged shouldBe (true)
            targetFileWithCurrentMetadata.hasChanged shouldBe (false)
            targetFileWithUpdatedCurrentGroup.hasChanged shouldBe (true)
            targetDirectoryWithoutCurrentMetadata.hasChanged shouldBe (true)
            targetDirectoryWithCurrentMetadata.hasChanged shouldBe (false)
            targetDirectoryWithUpdatedCurrentGroup.hasChanged shouldBe (true)
            targetLibraryWithoutCurrentMetadata.hasChanged shouldBe (true)
            targetLibraryWithCurrentMetadata.hasChanged shouldBe (false)
            targetLibraryWithUpdatedCurrentAttributes.hasChanged shouldBe (true)
        }

        "determine if its content has changed" {
            targetFileWithoutCurrentMetadata.hasContentChanged shouldBe (true)
            targetFileWithCurrentMetadata.hasContentChanged shouldBe (false)
            targetFileWithUpdatedCurrentSize.hasContentChanged shouldBe (true)
            targetFileWithUpdatedCurrentChecksum.hasContentChanged shouldBe (true)
            targetDirectoryWithoutCurrentMetadata.hasContentChanged shouldBe (false)
            targetDirectoryWithCurrentMetadata.hasContentChanged shouldBe (false)
            targetDirectoryWithUpdatedCurrentGroup.hasContentChanged shouldBe (false)
            targetLibraryWithoutCurrentMetadata.hasContentChanged shouldBe (true)
            targetLibraryWithCurrentMetadata.hasContentChanged shouldBe (false)
            targetLibraryWithUpdatedCurrentChecksum.hasContentChanged shouldBe (true)
        }

        "provide its original file path" {
            targetFile.originalRef.key shouldBe (targetFile.existingMetadata.path)
        }

        "provide its destination file path" {
            val testDestinationPath = Paths.get("/tmp/destination")
            val originalPath = Fixtures.Metadata.FileOneMetadata.path.asPath()

            val expectedPathWithDefaultStructure = Paths.get("$testDestinationPath/$originalPath")
            val expectedPathWithoutDefaultStructure = Paths.get("$testDestinationPath/${originalPath.fileName}")

            targetFile.destinationRef shouldBe (targetFile.originalRef)

            targetFile
                .copy(
                    destination = TargetEntity.Destination.Directory(
                        path = testDestinationPath,
                        keepDefaultStructure = true,
                        preserveExisting = false
                    )
                )
                .destinationRef shouldBe (EntityRef.Filesystem(expectedPathWithDefaultStructure))

            targetFile
                .copy(
                    destination = TargetEntity.Destination.Directory(
                        path = testDestinationPath,
                        keepDefaultStructure = false,
                        preserveExisting = false
                    )
                )
                .destinationRef shouldBe (EntityRef.Filesystem(expectedPathWithoutDefaultStructure))
        }

        "provide a filesystem destination for a library ref" {
            val testDestinationPath = Paths.get("/tmp/destination")

            val libraryFile = targetFile.copy(ref = EntityRef.Library(scheme = "photos", path = "/album/img.heic"))

            libraryFile
                .copy(
                    destination = TargetEntity.Destination.Directory(
                        path = testDestinationPath,
                        keepDefaultStructure = false,
                        preserveExisting = false
                    )
                )
                .destinationRef shouldBe (EntityRef.Filesystem(Paths.get("/tmp/destination/img.heic")))
        }

        "provide a non-colliding destination file path when preserving existing files" {
            val directory = Files.createTempDirectory("target-entity-preserve")
            directory.toFile().deleteOnExit()

            val fileName = Fixtures.Metadata.FileOneMetadata.path.asPath().fileName.toString()

            val entity = targetFile.copy(
                destination = TargetEntity.Destination.Directory(
                    path = directory,
                    keepDefaultStructure = false,
                    preserveExisting = true
                )
            )

            entity.destinationRef shouldBe (EntityRef.Filesystem(directory.resolve(fileName)))

            Files.createFile(directory.resolve(fileName))
            entity.copy().destinationRef shouldBe (EntityRef.Filesystem(directory.resolve("$fileName-1")))

            Files.createFile(directory.resolve("$fileName-1"))
            entity.copy().destinationRef shouldBe (EntityRef.Filesystem(directory.resolve("$fileName-2")))
        }
    }
})
